core/store/src/lib.rs

```diff
@@ -516,6 +516,38 @@ impl CompiledContractCache for StoreCompiledContractCache {
     }
 }
 
+/// Structure for fulfilling the `BlockHashProvider` interface using a DB look up.
+#[cfg(feature = "protocol_feature_block_hash_host_fn")]
+pub struct StoreBlockHashProvider {
+    store: Arc<Store>,
+    current_block_height: near_primitives::types::BlockHeight,
+}
+
+#[cfg(feature = "protocol_feature_block_hash_host_fn")]
+impl StoreBlockHashProvider {
+    pub fn new(
+        store: Arc<Store>,
+        current_block_height: near_primitives::types::BlockHeight,
+    ) -> Self {
+        Self { store, current_block_height }
+    }
+}
+
+#[cfg(feature = "protocol_feature_block_hash_host_fn")]
+impl near_primitives::types::BlockHashProvider for StoreBlockHashProvider {
+    fn block_hash(
+        &self,
+        height: near_primitives::types::BlockHeight,
+    ) -> Result<Option<CryptoHash>, std::io::Error> {
+        if height >= self.current_block_height
+            || height < self.current_block_height.saturating_sub(256)
```

Comment:
Maybe we need to have a named constant for 256

Reply:
Done in https://github.com/near/nearcore/pull/4273/commits/1c14806c9aeb06542c726d343ae83f369ccca553